### DISCUSSION TITLE
How to create two buttons, requiring SIWE and Not requiring SIWE  with RainbowKitSiweNextAuthProvider

### DISCUSSION THREAD

**Discussion by takadr** [rainbow-me/rainbowkit#1773] [Q&A] [answered]
Hi, I'm installing RainbowKitSiweNextAuthProvider to my project. 
I would like to create 2 kinds of buttons:

1. Requiring SIWE after connecting wallet (This type of users requires authentication to use the app because they interact with the DB)
2. Normal connect button, which doesn't require SIWE after connecting wallet (This type of users just interacts with the contract)

I guess it's possible if I use a boolean state in the store and pass it to the "enable" attribute of RainbowKitSiweNextAuthProvider, but is there a more elegant solution?

**Comment by 0xmkh** [comment, marked as answer]
@takadr That's correct you would need to pass `enable` true or false to `RainbowKitSiweNextAuthProvider`. I mean that's the approach i would go for. May i ask what is the end goal here ? What are you trying to achieve ?

**Reply by takadr** [reply]
Thanks for the answer.
My goal is putting 2 types of connect buttons at the same time for 2 types of users.
My image is using an atom and ConnectButton.Custom, and doing something like this:

Providers.tsx
``` js
const [requireAuth] = useAtom(requireAuthAtom);
...
<RainbowKitSiweNextAuthProvider enabled={requireAuth}>
  <RainbowKitProvider>{children}</RainbowKitProvider>
</RainbowKitSiweNextAuthProvider>
```

onClick in CustomConnectButton.tsx

Requiring SIWE:
```js
setRequireAuth(true);
openConnectModal();
```
Without SIWE:
```js
setRequireAuth(false);
openConnectModal();
```

Is this approach correct?
I was just wondering if there was a different approach like passing some configs to ConnectButton.

**Reply by 0xmkh** [reply]
This approach is correct yes. 

How about having two seperate RainbowKitProvider components one which has auth and one which doesn't have one.

Here is an example:

// RainbowKitAuthProvider.tsx
```
<RainbowKitAuthenticationProvider status={authStatus}>
    <RainbowKitProvider>
        // Internal auth logic here
        {children}
    </RainbowKitProvider>
</RainbowKitAuthenticationProvider>
```

// RainbowKitProvider.tsx
```
<RainbowKitProvider appInfo={demoAppInfo} chains={chains}>
    // Non auth logic here
    {children}
</RainbowKitProvider>
```
      
// App.tsx (main page)
`return requiredAuth ? <RainbowKitAuthProvider /> : <RainbowKitProvider />;`

Of course this is just an example, but i want you to decide the best approach. Let me know if this makes sense 🙏

**Reply by takadr** [reply]
Thanks 🙏 At least I achieved what I wanted to do with the code like below. (I needed to wait for the components to be re-rendered by putting setTimeout, which I think is a bit dirty)

```js
...
<button
  onClick={() => {
    setRequireAuth(false);
    setTimeout(() => {
      openConnectModal();
    }, 0);
  }}
  type="button"
>
  Connect Wallet
</button>
...
```

```js
...
<button
  onClick={() => {
    setRequireAuth(true);
    setTimeout(() => {
      openConnectModal();
    }, 0);
  }}
  type="button"
>
  Sign in with Ethereum
</button>
...
```

What I wanted to do is like this.
<img width="323" alt="スクリーンショット 2024-02-17 16 01 13" src="https://github.com/rainbow-me/rainbowkit/assets/92980001/ad827e34-69af-450a-bd4d-0c3b454047c0">

First button simply asks users to connect their wallet, while second one requests signing after connecting wallet.
Users can take different actions according to their signin status (not connected / connected / signed in).
So, I thought all components should be wrapped by just one AuthProvider.

My intuition was that I wanted to pass a config to disable the request for a signature after being connected, as an argument like

```js
<ConnectButton requireAuth={false}>
```
(This might not be consistent with the design philosophy though)

**Reply by 0xmkh** [reply]
Got it! 

The `ConnectButton` approach you mentioned would be kind of not super useful in the long term since not everyone uses that, but everyone uses `RainbowKitProvider` so maybe we can think of an approach to use that instead in the future. 

At least i like the approach you did and seems to work, but i've only got one question. Why do you need a timeout for opening a modal ? Can't you just open it right away ?

**Reply by takadr** [reply]
Sometimes it doesn't open up a modal. I guess it's because of the timing of re-rendering and I got it worked by adding setTimeout for now. Thanks anyway, now it's clear to me.
